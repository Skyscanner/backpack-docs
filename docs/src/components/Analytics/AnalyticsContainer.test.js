/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2019 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';

import AnalyticsContanier from './AnalyticsContainer';

jest.mock('../../helpers/google-analytics-helper', () => ({
  ...require.requireActual('../../helpers/google-analytics-helper'),
  startGoogleAnalytics: jest.fn(),
}));

jest.mock('../../helpers/localstorage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
}));

const localStorage = jest.requireMock('../../helpers/localstorage');
const ga = jest.requireMock('../../helpers/google-analytics-helper');
const storageKey = 'BPK_DOCS_accept-cookies';

describe('AnalyticsContainer', () => {
  afterEach(() => {
    ga.startGoogleAnalytics.mockClear();
    localStorage.getItem.mockClear();
    localStorage.setItem.mockClear();
  });

  it('renders correctly', () => {
    const tree = renderer.create(<AnalyticsContanier />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('rejects analytics when `reject` is clicked', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);

    ReactDOM.render(<AnalyticsContanier />, container);
    const rejectButton = container.querySelector(
      "[data-testid='reject-button']",
    );

    rejectButton.dispatchEvent(new MouseEvent('click', { bubbles: true }));

    expect(localStorage.setItem).toHaveBeenCalled();
    expect(localStorage.setItem.mock.calls[0][0]).toBe(storageKey);
    expect(localStorage.setItem.mock.calls[0][1]).toBe(false);
    expect(container.querySelector("[data-testid='cookie-banner']")).toBe(null);
  });

  it('accepts and starts analytics when `accept` is clicked', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);

    ReactDOM.render(<AnalyticsContanier />, container);
    const rejectButton = container.querySelector(
      "[data-testid='accept-button']",
    );

    rejectButton.dispatchEvent(new MouseEvent('click', { bubbles: true }));

    expect(localStorage.setItem).toHaveBeenCalled();
    expect(localStorage.setItem.mock.calls[0][0]).toBe(storageKey);
    expect(localStorage.setItem.mock.calls[0][1]).toBe(true);

    expect(ga.startGoogleAnalytics).toHaveBeenCalled();

    expect(container.querySelector("[data-testid='cookie-banner']")).toBe(null);
  });

  describe('when consent has been denied', () => {
    beforeEach(() => {
      localStorage.getItem.mockReturnValue('false');
    });

    it('renders correctly', () => {
      const tree = renderer.create(<AnalyticsContanier />).toJSON();

      expect(tree).toMatchSnapshot();
      expect(localStorage.getItem).toHaveBeenCalledWith(storageKey);
    });

    it('does not trigger startGoogleAnalytics', () => {
      renderer.create(<AnalyticsContanier />);

      expect(ga.startGoogleAnalytics).not.toHaveBeenCalled();
    });
  });

  describe('when consent has been granted', () => {
    beforeEach(() => {
      localStorage.getItem.mockReturnValue('true');
    });

    it('renders correctly', () => {
      const tree = renderer.create(<AnalyticsContanier />).toJSON();

      expect(tree).toMatchSnapshot();
      expect(localStorage.getItem).toHaveBeenCalledWith(storageKey);
    });

    it('triggers startGoogleAnalytics', () => {
      renderer.create(<AnalyticsContanier />);

      expect(ga.startGoogleAnalytics).toHaveBeenCalled();
    });
  });
});
