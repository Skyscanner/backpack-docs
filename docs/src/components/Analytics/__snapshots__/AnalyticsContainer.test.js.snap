// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnalyticsContainer renders correctly 1`] = `
<section
  className="bpkdocs-cookie-banner"
  data-testid="cookie-banner"
>
  <div
    className="bpkdocs-cookie-banner__inner"
  >
    <h2
      className="bpk-text bpk-text--xl"
      weight="heavy"
    >
      Your data. Your choice.
    </h2>
    <div
      className="bpkdocs-cookie-banner__message-wrapper"
    >
      <span
        className="bpk-text bpk-text--base"
      >
        Skyscanner uses Google Analytics on this site to help us analyse its use and improve our services. You can opt out by pressing "reject" or at any other time by using the
         
        <a
          className="bpk-link"
          href="https://tools.google.com/dlpage/gaoptout"
          onClick={null}
          rel={null}
          target="_blank"
        >
          Analytics opt-out add-on.
        </a>
      </span>
      <div
        className="bpkdocs-cookie-banner__buttons"
      >
        <button
          className="bpk-button"
          data-testid="accept-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Accept
        </button>
        <button
          className="bpk-button bpk-button--secondary"
          data-testid="reject-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Reject
        </button>
      </div>
    </div>
  </div>
</section>
`;

exports[`AnalyticsContainer when concent has been denied renders correctly 1`] = `null`;

exports[`AnalyticsContainer when concent has been granted renders correctly 1`] = `null`;
